typed/update_test.go

```diff
@@ -0,0 +1,141 @@
+package typed_test
+
+import (
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v4/internal/fixture"
+	"sigs.k8s.io/structured-merge-diff/v4/typed"
+)
+
+var updateParser = func() fixture.Parser {
+	parser, err := typed.NewParser(`
+types:
+- name: nestedOptionalFields
+  map:
+    fields:
+    - name: nestedList
+      type:
+        list:
+          elementRelationship: associative
+          keys:
+          - name
+          elementType:
+            map:
+                fields:
+                - name: name
+                  type:
+                    scalar: string
+                - name: value
+                  type:
+                    scalar: numeric                            
+    - name: nestedMap
+      type:
+        map:
+          elementType:
+            scalar: numeric
+    - name: nested
+      type:
+        map:
+          fields:
+            - name: numeric
+              type:
+                scalar: numeric
+            - name: string
+              type:
+                scalar: string
+`)
+	if err != nil {
+		panic(err)
+	}
+	return fixture.SameVersionParser{T: parser.Type("nestedOptionalFields")}
+}()
+
+func TestUpdate(t *testing.T) {
+	tests := map[string]fixture.TestCase{
+		"delete_nested_fields_struct": {
+			Ops: []fixture.Operation{
+				fixture.Apply{
+					Manager:    "default",
+					APIVersion: "v1",
+					Object: `
+                        nested:
+                            numeric: 1
+                            string: my string
+                    `,
+				},
+				fixture.Apply{
+					Manager:    "default",
+					APIVersion: "v1",
+					Object: `
+                        nested: {}
+                    `,
+				},
+			},
+			APIVersion: `v1`,
+			Object:     `{nested: {}}`,
+		},
+		"delete_nested_fields_list": {
+			Ops: []fixture.Operation{
+				fixture.Apply{
+					Manager:    "default",
+					APIVersion: "v1",
+					Object: `
+                        nestedList:
+                            - name: first
+                            - name: second
+                    `,
+				},
+				fixture.Apply{
+					Manager:    "default",
+					APIVersion: "v1",
+					Object: `
+                        nestedList: []
+                    `,
+				},
+			},
+			APIVersion: `v1`,
+			Object:     `{nestedList: []}`,
+		},
+		"delete_nested_fields_map": {
+			Ops: []fixture.Operation{
+				fixture.Apply{
+					Manager:    "default",
+					APIVersion: "v1",
+					Object: `
+                        nestedMap:
+                            first: 1
+                            second: 2
+                            
+                    `,
+				},
+				fixture.Apply{
+					Manager:    "default",
+					APIVersion: "v1",
+					Object: `
+                        nestedMap: {}
+                    `,
+				},
+			},
+			APIVersion: `v1`,
+			Object:     `{nestedMap: {}}`,
+		},
```

Comment:
Yeah, so I guess now I'm confused about:
1. What happens if we apply null instead.
2. What happens if you have further level of nesting (i.e. you're apply `{}` when you had `{'nested': {'numeric':1}}` (I'm guessing the same but would love to see)
3. How do you entirely remove the field? If null keeps null and {} keep it to empty, then how do you remove it altogether?

Reply:
Will add cases for 1 & 2

Re: 3. If you wanted to make `{nestedMap: { myField: value }}` into `{}` you would just apply `{}`. If you applied `{}` then `nestedMap` field is included in the `toRemove` set. If you had applied `{nestedMap: {}}` then only `nestedMap.myField` is in the `toRemove` set, so `nestedMap` should be retained

Reply:
For the null case naively I would expect it to keep null if you specified null.